server/src/main/java/org/elasticsearch/search/SearchService.java

```diff
@@ -1248,6 +1251,18 @@ private CanMatchResponse canMatch(ShardSearchRequest request, boolean checkRefre
                 if (canMatch || hasRefreshPending) {
                     FieldSortBuilder sortBuilder = FieldSortBuilder.getPrimaryFieldSortOrNull(request.source());
                     minMax = sortBuilder != null ? FieldSortBuilder.getMinMaxOrNull(context, sortBuilder) : null;
+
+                    if (RollupV2.isEnabled()) {
+                        IndexMetadata requestIndexMetadata = clusterService.state().getMetadata()
+                                .index(request.shardId().getIndexName());
+
+                        CanMatchResponse rollupCanMatchResponse = RollupShardDecider.canMatch(request, context, requestIndexMetadata,
```

Comment:
I wonder if that would be easier to not copy the `RollupShardDecider` from v1. That's not how we want to implement the check so no need to copy tests and stuffs since we'll need to rewrite them anyway ?
We can start simple and only check if the original index pattern (`ShardSearchRequest#originalIndices`)  contains the datastream of the rollup index. Then we can incrementally add the verification inside the mapping, queries and agg, we don't need to get it right on the first PR.